Add basic setup switching for the tournament client

MiraiSubject opened the pull request:
- [x] Depends on https://github.com/ppy/osu/pull/9247 for tests to pass

From https://github.com/ppy/osu/issues/9199, this fulfils 
> Create a default tournament profile and the necessary migrations for the flags, mods, videos and other assets.

This PR changes the default path for tournaments from `%APPDATA%\osu\tournament`  to `%APPDATA%\osu\tournaments\default`.
Migrations for this are implemented so that this change does not break existing configurations and the necessary tests are also written for that. The migration also includes moving the `bracket.json` to the same directory. 

Upon migration a `tournament.ini` gets created with the following default value:
```
CurrentTournament = default
```
## Post-migration scenario
The user can now add directories in the following format:
```
(%APPDATA%/osu/tournaments)
tournament-name
 ┃ ┣ config
 ┃ ┃ ┗ drawings.ini
 ┃ ┣ flags
 ┃ ┣ mods
 ┃ ┣ videos
 ┃ ┣ bracket.json
 ┃ ┣ drawings.txt
 ┃ ┗ drawings_results.txt
```
and then change the `tournament.ini` to:
```
CurrentTournament = tournament-name
```
Because of this, switching the storage requires a reboot of the tournament client.

MiraiSubject committed:
Introduce new storage class and manager

MiraiSubject committed:
Rename to TournamentStorage

MiraiSubject committed:
Implement migration for TournamentStorage

MiraiSubject committed:
Fix styling issues and move StorageManager to Configuration Folder

MiraiSubject committed:
Refactor paths

MiraiSubject committed:
Remove unnecessary parameters and implement delete

MiraiSubject committed:
Code styling fixes

MiraiSubject committed:
change namespace to osu.Game.Tournament.IO

MiraiSubject committed:
Rename tournamentStorage to storage

MiraiSubject committed:
Fixed migration logic

MiraiSubject committed:
Add NonVisual tests for custom tournaments

Can test the default directory from a clean instance, it can test a custom directory and can execute migration from an instance using the older directory setup.

MiraiSubject committed:
Remove misleading log

MiraiSubject committed:
Change access modifier public -> internal

MiraiSubject committed:
Rename VideoStorage to VideoStore

MiraiSubject committed:
move TournamentVideoResourceStore to separate file

MiraiSubject committed:
Add a comment to describe what's going on before the headless game starts

MiraiSubject committed:
Remove unused property and reuse tournamentBasePath

MiraiSubject committed:
Add newlines at the end of the file

MiraiSubject committed:
Undo blank line removal

Was too excited to add blank lines before submitting the PR that I overdid it

MiraiSubject committed:
Introduce migrations for drawings

MiraiSubject committed:
Switch drawing storage to tournamentstorage

MiraiSubject committed:
Create storage for config files of a tournament

MiraiSubject committed:
Implement migrations for the drawings config file

MiraiSubject committed:
Add drawings to the migration test

peppy commented:
Generally if something depends on another branch/PR, you should rebase it on top of that (so it will compile and work for testing).

MiraiSubject commented:
> Generally if something depends on another branch/PR, you should rebase it on top of that (so it will compile and work for testing).

I locally tested by merging this branch into https://github.com/ppy/osu/pull/9247, stashed any changes I wanted to keep and then applied it to this branch. It's a very roundabout way of doing things, but it was the only way I was familiar and comfortable with. 

I'll look into rebasing.

peppy commented:
It's fine this time around, just a heads-up for future contributions.

peppy committed:
Merge branch 'master' into tourney-asset-refactor

peppy committed:
Fix merge issue

peppy commented on osu.Game.Tournament/IO/TournamentStorage.cs:
```diff
@@ -0,0 +1,153 @@
+        private const string default_tournament = "default";
+        private const string config_directory = "config";
+
+        public TournamentStorage(GameHost host)
```
Why can't this just take in a storage, rather than the host itself?

peppy commented on osu.Game.Tournament/IO/TournamentStorage.cs:
```diff
@@ -0,0 +1,153 @@
+            if (!cfgDestination.Exists)
+                destination.CreateSubdirectory(config_directory);
+
+            if (host.Storage.Exists("bracket.json"))
```
please learn how to avoid doing this by using functions.

peppy commented on osu.Game.Tournament/IO/TournamentStorage.cs:
```diff
@@ -0,0 +1,153 @@
+                moveFile(bracketFile, destination);
+            }
+
+            if (host.Storage.Exists("drawings.txt"))
```
why can't you use the existing `Migrate` method? this is mostly copy-paste of existing implementation.

peppy commented on osu.Game.Tournament/IO/TournamentStorage.cs:
```diff
@@ -0,0 +1,153 @@
+            }
+        }
+
+        private void copyRecursive(DirectoryInfo source, DirectoryInfo destination)
```
don't copy paste please.

peppy commented on osu.Game.Tournament/IO/TournamentStorage.cs:
```diff
@@ -0,0 +1,153 @@
+        internal readonly TournamentVideoResourceStore VideoStore;
+        internal readonly Storage ConfigurationStorage;
+        private const string default_tournament = "default";
+        private const string config_directory = "config";
```
why even bother? isn't there only one config file ever?

MiraiSubject committed:
TournamentStorage now takes in a parent storage

MiraiSubject committed:
Refactor moving logic (1/2)

MiraiSubject committed:
Refactor moving (2/2)

MiraiSubject committed:
Introduce new class MigratableStorage

MiraiSubject committed:
Move static properties to parent class and inherit OsuStorage from it

MiraiSubject committed:
Fix TestMigrationToSeeminglyNestedTarget failing

MiraiSubject committed:
Merge branch 'master' of https://github.com/ppy/osu into tourney-asset-refactor

MiraiSubject committed:
Remove unnecessary comments and added file check for tournament.ini on test start

MiraiSubject committed:
Move drawings.ini out of config subfolder

MiraiSubject committed:
InspectCode changes

MiraiSubject committed:
Add xmldoc to MigratableStorage

MiraiSubject committed:
make tournament migration private

MiraiSubject committed:
Remove whitespace at the end of xmldoc line

MiraiSubject commented:
There a couple of changes I made to this: 

- Introduce a new abstract class `MigratableStorage` that inherits from `WrappedStorage`
- Change the inheritance for `OsuStorage` and `TournamentStorage` to `MigratableStorage`
- `TournamentStorage` now takes in an existing storage as ctor parameter
- File moving in `TournamentStorage` is now called `moveFileIfExists` to prevent filling up `migrate()` up with if-statements.
- `drawings.ini` has been moved back from `tournaments/default/config` to `tournaments/default`

`MigratableStorage` now contains the Copy, Delete and Attempt operation methods as well as the constants for `ignore_files` and `ignore_directories`

I also made sure all the tests for `CustomDataDirectoryTest` still pass. 

I'm not sure if the `ignore_files` and `ignore_directories` should stay there, because imo every `MigratableStorage` should have their own array of ignored files and directories and now it's globally declared across every subclass. I just haven't found a solution to this yet that does not look bad. If anybody has an idea about this, let me know.

bdach commented on osu.Game/IO/MigratableStorage.cs:
```diff
@@ -0,0 +1,99 @@
+    /// </summary>
+    public abstract class MigratableStorage : WrappedStorage
+    {
+        internal static readonly string[] IGNORE_DIRECTORIES = { "cache" };
+
+        internal static readonly string[] IGNORE_FILES =
+        {
+            "framework.ini",
+            "storage.ini"
+        };
```
The obvious way to make this customisable by consumers is by making these into protected get-only virtual properties.

bdach commented on osu.Game.Tournament/Components/TourneyVideo.cs:
```diff
@@ -29,7 +30,7 @@ public TourneyVideo(string filename, bool drawFallbackGradient = false)
         private void load(TournamentStorage storage)
         {
-            var stream = storage.GetStream($@"videos/{filename}");
+            var stream = storage.VideoStore.GetStream($@"{filename}");
```
This string interpolation is not required any more.
```suggestion
            var stream = storage.VideoStore.GetStream(filename);
```

bdach commented on osu.Game.Tournament/Configuration/TournamentStorageManager.cs:
```diff
@@ -0,0 +1,29 @@
+        protected override void InitialiseDefaults()
+        {
+            base.InitialiseDefaults();
+            Set(StorageConfig.CurrentTournament, string.Empty);
```
Is this even necessary if it's going to be overwritten upon first execution of `TournamentStorage` anyway?

bdach commented on osu.Game.Tournament/IO/TournamentStorage.cs:
```diff
@@ -0,0 +1,72 @@
+    public class TournamentStorage : MigratableStorage
+    {
+        private readonly Storage storage;
+        internal readonly TournamentVideoResourceStore VideoStore;
```
The `internal` looks really strange to me. I'd write this as

```suggestion
        public readonly TournamentVideoResourceStore VideoStore { get; }
```

bdach commented on osu.Game.Tournament/Screens/Drawings/Components/StorageBackedTeamList.cs:
```diff
@@ -14,9 +14,9 @@ public class StorageBackedTeamList : ITeamList
 
-        private readonly Storage storage;
+        private readonly TournamentStorage storage;
 
-        public StorageBackedTeamList(Storage storage)
+        public StorageBackedTeamList(TournamentStorage storage)
```
Is this required? A `TournamentStorage` is a `Storage` still...

bdach commented on osu.Game.Tournament/Screens/Drawings/DrawingsScreen.cs:
```diff
@@ -36,12 +36,12 @@ public class DrawingsScreen : TournamentScreen
 
         [BackgroundDependencyLoader]
-        private void load(TextureStore textures, Storage storage)
+        private void load(TextureStore textures, TournamentStorage storage)
```
This one is confusing too, but I can understand why this would be required more (due to caching in `TournamentGameBase`, I presume). You can `CacheAs<Storage>`, fwiw.

MiraiSubject committed:
make ignore properties protected virtual get-only in base

MiraiSubject committed:
Remove default value in Storagemgr

MiraiSubject committed:
make migrate public abstract in base and override

MiraiSubject committed:
invert and early return

MiraiSubject committed:
Remove string interpolation & unnecessary test setup

MiraiSubject committed:
change accessor from internal readonly to public get-only

Also changes the class accessor from internal to public

MiraiSubject committed:
Cache TournamentStorage as Storage and only cast when necessary

MiraiSubject committed:
Address formatting issues

MiraiSubject committed:
Fix nullref exceptions and redundant explicit type

MiraiSubject committed:
Change type from TournamentStorage to Storage in tests

MiraiSubject committed:
Check if the file exists before reading

This is (also) to address the review from bdach about StorageManager initialising a default value that gets overwritten upon migration anyway.

MiraiSubject commented on osu.Game.Tests/NonVisual/CustomDataDirectoryTest.cs:
```diff
@@ -151,13 +152,13 @@ public void TestMigration()
                     }
 
-                    foreach (var dir in OsuStorage.IGNORE_DIRECTORIES)
+                    foreach (var dir in osuStorage?.IgnoreDirectories ?? Array.Empty<string>())
```
Not sure if this is the right way to prevent a Nullref exception and it also looks weird to loop over an empty array

peppy replied:
You're correct in thinking this is not good code. just do a null check *before* iterating instead (you can use it to surround both the two `foreach` statements).

MiraiSubject committed:
Move null check outside of the loop

bdach commented on osu.Game.Tests/NonVisual/CustomDataDirectoryTest.cs:
```diff
@@ -151,16 +152,19 @@ public void TestMigration()
                     Assert.That(storage.ExistsDirectory(Path.Combine("test-nested", "cache")));
 
-                    foreach (var file in OsuStorage.IGNORE_FILES)
+                    if (osuStorage != null)
```
Shouldn't this be an actual assertion, as in the following?

```csharp
Assert.That(osuStorage, Is.Not.Null);
```

Imagine this regresses for some reason and `osuStorage` becomes null; this `if` will then mean that the test will silently skip over the nested asserts and stop testing if the ignore lists work at all.

(As an aside, I think it'd also make sense to change type of `osuStorage` to `MigratableStorage` as it's the most general type possible in this scenario)

MiraiSubject replied:
Good point! I'll add that and change the cast right away. I'll also include the property accessor change right away.

bdach commented on osu.Game/IO/MigratableStorage.cs:
```diff
@@ -0,0 +1,96 @@
+    /// </summary>
+    public abstract class MigratableStorage : WrappedStorage
+    {
+        internal virtual string[] IgnoreDirectories => Array.Empty<string>();
+        internal virtual string[] IgnoreFiles => Array.Empty<string>();
```
This is a weird modifier mix. Any consumers from a different assembly that might want to override this will not be able to. I think this calls for the [rarely-seen](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/protected-internal)

```suggestion
        protected internal virtual string[] IgnoreDirectories => Array.Empty<string>();
        protected internal virtual string[] IgnoreFiles => Array.Empty<string>();
```

(Or heck, just make it public, I don't think that's that much of a deal.)

Needs to be fixed up in the override too if this happens.

MiraiSubject replied:
I've changed it to public for now, will await further feedback about this in case `protected` becomes the preferred way to go.

MiraiSubject committed:
Address review comments

Now asserting instead of an if-statement, change cast from OsuStorage to MigratableStorage and make internal virtual properties protected.

MiraiSubject committed:
Reuse custom_tournament where it was still used as a literal

MiraiSubject committed:
Merge branch 'master' into tourney-asset-refactor

MiraiSubject committed:
Move storageconfig set and saving to migrate method

MiraiSubject committed:
Move general purpose migration to MigratableStorage

MiraiSubject commented:
I've been looking into making `Migrate` more sane.

Right now I moved the original `Migrate` from `OsuStorage` to `MigratableStorage` (+xmldoc) so that consumers can use this as a general purpose migrator.

I've also looked into making the `TournamentStorage`'s `Migrate` use `MigratableStorage`'s but that's not possible because the destination is nested inside the source. That goes along with the structure of what's being migrated is different from what it would be in a general case. 

Before:
```
(%APPDATA%/osu/)
... cache, files, etc.
 ┃ ┣ tournament/
 ┃ ┃ ┣  flags
 ┃ ┃ ┣  mods
 ┃ ┃ ┗  videos
 ┃ ┣  bracket.json
 ┃ ┣ drawings.txt.
 ┃ ┣ drawings_results.txt.
 ┃ ┗ drawings.ini
...  framework.ini, online.db, etc.
```

After: 
```
(%APPDATA%/osu/tournaments)
default
 ┃ ┣ flags
 ┃ ┣ mods
 ┃ ┣ videos
 ┃ ┣ bracket.json
 ┃ ┣ drawings.txt
 ┃ ┣ drawings.ini
 ┃ ┗ drawings_results.txt
```

bdach commented:
As a heads up, but this is probably going to have major merge conflicts with #9417 / #9418...

As for the issue mentioned I'd probably advise that the `Migrate()` method be overridable and that `TournamentStorage` hoist the files one level out of `tournament/` as a post-migration step.

MiraiSubject committed:
Logger now shows the actual path of the destination

Forgot to change this while changing the param from string to Storage

MiraiSubject commented:
Alright thanks! I'll hold off from any more big changes until after those are merged

I made `Migrate` already overridable and its implementation in `TournamentStorage` is being completely overridden without ever calling `base` because this exception in `MigratableStorage` will get thrown:

```csharp
if (sourceUri.IsBaseOf(destinationUri))
    throw new ArgumentException("Destination provided is inside the source", destination.FullName);
```

I also believe at the same time that `TournamentStorage`'s `Migrate` is different enough to not need to call `base`.

I could just be talking about some non-issue and that overriding like this is fine, but I am not sure. Hopefully it will come to light once it can be reviewed again.

MiraiSubject committed:
Merge branch 'master' into tourney-asset-refactor

# Conflicts:
#	osu.Game/IO/OsuStorage.cs

MiraiSubject commented:
I resolved the conflicts in an incoming commit and reran the tests locally again to make sure they didn't fail. 

I thought some more about this and my comment above can actually be disregarded. The intention of the migration method is that it is supposed to be a temporary method so that existing configurations get migrated after users update the client.

For future versions users would be instructed to create their new configurations in the format of `tournaments/<tournament acronym>` and then edit the `tournament.ini` appropriately to `CurrentTournament = <tournament acronym>`. 

The end goal is to put this functionality of adding and selecting tournaments in the UI itself.

peppy committed:
Merge branch 'master' into tourney-asset-refactor

peppy committed:
Merge branch 'master' into tourney-asset-refactor

peppy committed:
Add comment about migration code

peppy committed:
Update tests to use correct parameters of CleanRunGameHost

peppy committed:
Add missing xmldoc

peppy committed:
Move video store out of TournamentStorage

There was no reason it should be nested inside.

peppy committed:
Fix texture store not being initialised correctly

Without this change flags/mods would not work as expected. The video
store was being added as the texture store incorrectly.

peppy reviewed (approved):
Should be good to go now.
